roachtestsqlsmith: release-22.1: internal error: no output column equivalent

roachtest.sqlsmith/setup=rand-tables/setting=no-mutations [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=5645973&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=5645973&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations) on release-22.1 @ [143d142bd231894a93813c75607702156db429ae](https://github.com/cockroachdb/cockroach/commits/143d142bd231894a93813c75607702156db429ae):


```
The test failed on branch=release-22.1, cloud=gce:
test artifacts and logs in: /artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:265,sqlsmith.go:324,test_runner.go:883: error: pq: internal error: no output column equivalent to 9
		stmt:
		WITH
			with_58563 (col_292367)
				AS (
					SELECT NULL AS col_292367 FROM defaultdb.public.table2@[0] AS tab_156225
					INTERSECT
						SELECT
							*
						FROM
							(
								VALUES
									(
										e'{"$qoz3v": [{"b": [[]], "bar": {"PhwGlG": [], "c": []}}, []], "\'D- |I2^+[3?": null, ";Lb#8AP->Y": {"b": [[]]}, "x|+G*p$81tXj": [{}]}':::JSONB
									),
									(NULL),
									(
										e'{"o._{;X": 0.19412523575721738, "v\\"_0": [{"3:7$)7PI)": "bar", "b": {"[IFvuRsrm)": [], "a": {"u/Qg%Y\\\\:e2>Z": []}}, "uz#PL%NEiO8j": {}}, [[], []], {"foo": {"HCtjw``m4hqZQ": "baz"}}]}':::JSONB
									),
									(NULL),
									('0.5149353168111566':::JSONB)
							)
								AS tab_156226 (col_292368)
				)
		SELECT
			count(*) AS col_292369
		FROM
			defaultdb.public.table1@[0] AS tab_156227
		WHERE
			(NOT (tab_156227.col1_8 NOT SIMILAR TO e'Eb\nhc]\x7f\x03':::STRING))
		GROUP BY
			tab_156227.col1_7
		HAVING
			bool_and(true::BOOL)::BOOL
		LIMIT
			10:::INT8;
```
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #83791 roachtest: sqlsmith/setup=rand-tables/setting=no-mutations failed [C-test-failure O-roachtest O-robot T-sql-queries branch-master]
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-mutations.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-17269

--- a/pkg/sql/opt/ordering/scan.go
+++ b/pkg/sql/opt/ordering/scan.go
@@ -130,14 +130,20 @@ func scanBuildProvided(expr memo.RelExpr, required *props.OrderingChoice) opt.Or
 	// We generate the longest ordering that this scan can provide, then we trim
 	// it. This is the longest prefix of index columns that are output by the scan
 	// (ignoring constant columns, in the case of constrained scans).
-	// We start the for loop at the exact prefix since all columns in the exact
-	// prefix are constant and can be ignored.
+	outCols := expr.Relational().OutputCols
 	constCols := fds.ComputeClosure(opt.ColSet{})
 	numCols := index.KeyColumnCount()
 	provided := make(opt.Ordering, 0, numCols)
-	for i := scan.ExactPrefix; i < numCols; i++ {
+	for i := 0; i < numCols; i++ {
 		indexCol := index.Column(i)
 		colID := scan.Table.ColumnID(indexCol.Ordinal())
+		if i < scan.ExactPrefix && !outCols.Contains(colID) {
+			// All columns in the exact prefix are constant and can be ignored as long
+			// as they are not in the output of the scan. If an exact-prefix column is
+			// in the output, it may still be constant, but can only be ignored if the
+			// FDs "know" it is constant. This latter case is handled below as normal.
+			continue
+		}
 		if constCols.Contains(colID) {
 			// Column constrained to a constant, ignore.
 			continue

--- a/pkg/sql/opt/xform/testdata/physprops/ordering
+++ b/pkg/sql/opt/xform/testdata/physprops/ordering
@@ -2899,3 +2899,35 @@ project
                 ├── t0_85393.rowid:10 [as=t0_85393.rowid:2, outer=(10)]
                 ├── t1_85393.c0:13 [as=t1_85393.c0:5, outer=(13)]
                 └── t1_85393.rowid:14 [as=t1_85393.rowid:6, outer=(14)]
+
+# Regression test for #83793 - include scan columns that are constrained to be
+# constant in the provided ordering when they are in the output of the scan.
+exec-ddl
+CREATE TABLE t83793 (
+  a INT,
+  b STRING AS (a::STRING) STORED,
+  c STRING AS (a::STRING) VIRTUAL,
+  UNIQUE (b, a)
+);
+----
+
+opt format=hide-all
+SELECT NULL FROM t83793
+WHERE NOT (c NOT SIMILAR TO '')
+GROUP BY b HAVING bool_and(NULL);
+----
+project
+ ├── select
+ │    ├── group-by (streaming)
+ │    │    ├── project
+ │    │    │    ├── scan t83793@t83793_b_a_key
+ │    │    │    │    └── constraint: /2/1: [/'' - /'']
+ │    │    │    └── projections
+ │    │    │         └── NULL
+ │    │    └── aggregations
+ │    │         └── bool-and
+ │    │              └── column7
+ │    └── filters
+ │         └── bool_and
+ └── projections
+      └── NULL

